### ISSUE TITLE
DEBUG: .../IO/Socket/SSL.pm:1177: global error: Undefined SSL object

### ISSUE THREAD

**Issue by mendel5** [imapsync/imapsync#293]
I receive a debugging statement when working with imapsync. My imapsync is running on a Linux Mint 20.1 system.

Here is my command (partially adjusted to not reveal my credentials):
`
./imapsync --host1 'imap.old-hosting-provider.com' --port1 993 --user1 'old.email@mydomain1.com' --password1 'mypassword1' --host2 'imap.new-hosting-provider.com' --port2 993 --user2 'new.email@mydomain2.com' --password2 'mypassword2' --ssl1 --sslargs1 SSL_verify_mode=1 --sslargs1 SSL_version=TLSv1_2 --ssl2 --sslargs2 SSL_verify_mode=1 --sslargs2 SSL_version=TLSv1_3 --expunge1 --subscribeall
`

The debug statement:
```
DEBUG: .../IO/Socket/SSL.pm:1177: global error: Undefined SSL object
```
This debug message appears 4 times, twice for the connection to host1 and twice for the connection to host2.

However, the TLS encryption for both connections - TLS1.2 to host1 and TLS1.3 to host2 - works correctly. I have checked that by using Wireshark to inspect my own traffic. The transfer of the emails from host1 to host2 finishes successfully despite the error message.

More details:
```
imapsync version 1.977

Modules version list:
Authen::NTLM         1.09
CGI                  4.46
Compress::Zlib       2.093
Crypt::OpenSSL::RSA  0.31
Data::Uniqid         0.12
Digest::HMAC_MD5     1.01
Digest::HMAC_SHA1    1.03
Digest::MD5          2.55
Encode               3.01
Encode::IMAPUTF7     1.05
File::Copy::Recursive 0.45
File::Spec           3.78
Getopt::Long         2.5
HTML::Entities       3.69
IO::Socket           1.40
IO::Socket::INET     1.40
IO::Socket::INET6    2.72
IO::Socket::IP       0.39
IO::Socket::SSL      2.067
IO::Tee              0.65
JSON                 4.02
JSON::WebToken       0.10
LWP                  6.43
MIME::Base64         3.15
Mail::IMAPClient     3.43
Net::Ping            2.71
Net::SSLeay          1.88
Term::ReadKey        2.38
Test::MockObject     1.20200122
Time::HiRes          1.976
URI::Escape          3.31
Unicode::String      2.10

SSL debug mode level is --debugssl 1 (can be set from 0 meaning no debug to 4 meaning max debug)
```

If you need any other info please let me know.

**Comment by gilleslamiral** [comment]
Do you have a real concrete example I can reproduce? Something like:

    imapsync --host1 'test1.lamiral.info'  --sslargs1 SSL_verify_mode=1 --sslargs1 SSL_version=TLSv1_2 

but with an other host than test1.lamiral.info, a one that fails and I can play with.

**Comment by mendel5** [comment]
I tried to reproduce the issue with the following command:

```
./imapsync --host1 'imap.old-hosting-provider.com' --ssl1 --sslargs1 SSL_verify_mode=1 --sslargs1 SSL_version=TLSv1_2
```

Returns:
```
Exiting after a justconnect on host(s): imap.old-hosting-provider.com
Exiting with return value 0 (EX_OK: successful termination) 0/50 nb_errors/max_errors
```

It seems that the issue cannot be reproduced by using only this command. An actual login to a mailbox including the transfer of e-mails might be necessary to reproduce the debug message.

**Comment by gilleslamiral** [comment]
> I tried to reproduce the issue with the following command:
> 
> ```
> ./imapsync --host1 'imap.old-hosting-provider.com' --ssl1 --sslargs1 SSL_verify_mode=1 --sslargs1 SSL_version=TLSv1_2
> ```
> 
> Returns:
> 
> ```
> Exiting after a justconnect on host(s): imap.old-hosting-provider.com
> Exiting with return value 0 (EX_OK: successful termination) 0/50 nb_errors/max_errors
> ```

Good!

> There is no debug statement. It seems that the issue cannot be reproduced by using only this command. 

Debug options:

    imapsync ... --debugssl=4 --debugimap


> An actual login to a mailbox including the transfer of e-mails might be necessary to reproduce the debug message.


You're right, I can not reproduce the error myself, but I get another one: imap.old-hosting-provider.com doesn't resolve at all.
Can you give a real concrete example I can reproduce, Is imap.old-hosting-provider.com the real hostname?

**Comment by mendel5** [comment]
> Debug options:
> ```
> imapsync ... --debugssl=4 --debugimap
> ```

I tried to add it to the previous command but the error message can still not be reproduced.

> Is imap.old-hosting-provider.com the real hostname?

No, it's not the real hostname. I just used it as an example.

> Can you give a real concrete example I can reproduce

Yes. Can I contact you via email at `gilles@lamiral.info`?

**Comment by gilleslamiral** [comment]
Sure!

**Comment by mendel5** [comment]
The debug statement
```
DEBUG: .../IO/Socket/SSL.pm:1177: global error: Undefined SSL object
```
originates from the Perl module `IO::Socket::SSL`.

As can be seen in the first comment, the version of `IO::Socket::SSL` that has been used is `2.067`. At the time of writing this comment, the current version is `2.071`. Maybe the debug message has been fixed in the meantime.

Links to `IO::Socket::SSL`:
- https://metacpan.org/pod/IO::Socket::SSL
- https://metacpan.org/dist/IO-Socket-SSL/changes

Since the debug message does not orignate from `imapsync` but from `IO::Socket::SSL`, I will close this issue.

**Comment by OldGrumpy-de** [comment]
Unfortunately, the corresponding issue over at IO::Socket::SSL has been closed as well, because they argue there is something wrong in the caller program. Link: [https://github.com/noxxi/p5-io-socket-ssl/issues/109]

I happened to find this thread because it occurs randomly during transfers. I could not find any pattern, just that it always appears twice. I can create a detailed log for you, just let me know what debug options to set.

**Comment by mendel5** [comment]
> I can create a detailed log for you, just let me know what debug options to set.

Hi @OldGrumpy-de , as far as I know the issue does not originate from imapsync, so gilleslamiral cannot do much about it.

Maybe we can reopen the issue at https://github.com/noxxi/p5-io-socket-ssl/issues/109

**Comment by noxxi** [comment]
It looks for me like the method blocking with called on an object which is not yet or no longer an SSL object. But that's all I can get from the error message. It is unlikely that the problem is caused by imapsync directly though since imapclient does not use IO::Socket::SSL directly.

Unfortunately it looks like there are not enough information for others to reproduce the problem, which make debugging and fixing it a problem. The severity seems to be low though, i.e. just an error message without any reported side effects.

**Comment by gilleslamiral** [comment]
I'll fix it when I'll encounter it reproductively with a bad side effect.

**Comment by mendel5** [comment]
All right, thanks for the feedback from both of you.

**Comment by mildred** [comment]
I have the same error (DEBUG: .../IO/Socket/SSL.pm:1177: global error: Undefined SSL object​) but imapsync is executed without any ssl specific flag. It didn't occur before. Also, it sems either very slow to sync, or it is stalling on the INBOX folder (running a partial sync with most of the mails synchronized before). The message just appear many times as in a loop and the sync does not progress.

I'll try to provide more useful logs with `--debugssl=4 --debugimap`

**Comment by mildred** [comment]
It seems caused by a stalled IMAP connection (at least for me)...

```
++++ Verifying [INBOX] -> [INBOX]
Sending: 4068 STATUS INBOX (UIDNEXT)
Sent 29 bytes
Read:   * STATUS INBOX (UIDNEXT 4142)
        4068 OK Completed
Host2: uidnext is 4142
Host1: found that msg INBOX/9582 equals Host2 INBOX/664
Host1: flags init msg INBOX/9582 flags( \Seen NonJunk category.Perso ) Host2 msg INBOX/664 flags( \Seen )
Host1: flags filt msg INBOX/9582 flags( \Seen ) Host2 msg INBOX/664 flags( \Seen )
Host1: size msg INBOX/9582 = 3140 <> 3140 = Host2 INBOX/664
Host1: found that msg INBOX/9583 equals Host2 INBOX/665
Host1: flags init msg INBOX/9583 flags( \Seen NonJunk category.Perso category.Perso.remy20 ) Host2 msg INBOX/665 flags( \Seen )
Host1: flags filt msg INBOX/9583 flags( \Seen ) Host2 msg INBOX/665 flags( \Seen )
Host1: size msg INBOX/9583 = 37381 <> 37381 = Host2 INBOX/665
Host1: found that msg INBOX/9584 equals Host2 INBOX/666
Host1: flags init msg INBOX/9584 flags( \Seen NonJunk category.Perso ) Host2 msg INBOX/666 flags( \Seen )
Host1: flags filt msg INBOX/9584 flags( \Seen ) Host2 msg INBOX/666 flags( \Seen )
Host1: size msg INBOX/9584 = 7038 <> 7038 = Host2 INBOX/666
Host1: found that msg INBOX/9585 equals Host2 INBOX/667
Host1: flags init msg INBOX/9585 flags( \Seen ) Host2 msg INBOX/667 flags( \Seen )
Host1: flags filt msg INBOX/9585 flags( \Seen ) Host2 msg INBOX/667 flags( \Seen )
Host1: size msg INBOX/9585 = 8403 <> 8403 = Host2 INBOX/667
Host1: found that msg INBOX/9586 equals Host2 INBOX/668
Host1: flags init msg INBOX/9586 flags( NonJunk category.Perso ) Host2 msg INBOX/668 flags( NonJunk )
Host1: flags filt msg INBOX/9586 flags(  ) Host2 msg INBOX/668 flags( NonJunk )
Host2: flags msg INBOX/668 replacing h2 flags( NonJunk ) with h1 flags(  )
Sending: 4069 UID STORE 668 FLAGS.SILENT ()
Sent 36 bytes
ERROR: timeout waiting 120s for data from server at /usr/share/perl5/Mail/IMAPClient.pm line 1893.
        Mail::IMAPClient::__read_more(Mail::IMAPClient=HASH(0x5564453014d8), IO::Socket::SSL=GLOB(0x556445302250), 120) called at /usr/share/perl5/Mail/IMAPClient.pm li
ne 1859
        Mail::IMAPClient::_read_more(Mail::IMAPClient=HASH(0x5564453014d8), IO::Socket::SSL=GLOB(0x556445302250), 120) called at /usr/share/perl5/Mail/IMAPClient.pm lin
e 1670
        Mail::IMAPClient::_read_line(Mail::IMAPClient=HASH(0x5564453014d8)) called at /usr/share/perl5/Mail/IMAPClient.pm line 1406
        Mail::IMAPClient::_get_response(Mail::IMAPClient=HASH(0x5564453014d8), 4069, undef) called at /usr/share/perl5/Mail/IMAPClient.pm line 1340
        Mail::IMAPClient::_imap_command_do(Mail::IMAPClient=HASH(0x5564453014d8), "UID STORE 668 FLAGS.SILENT ()") called at /usr/share/perl5/Mail/IMAPClient.pm line 12
38
        Mail::IMAPClient::_imap_command(Mail::IMAPClient=HASH(0x5564453014d8), "UID STORE 668 FLAGS.SILENT ()") called at /usr/share/perl5/Mail/IMAPClient.pm line 1450
        Mail::IMAPClient::_imap_uid_command(Mail::IMAPClient=HASH(0x5564453014d8), "STORE", 668, "FLAGS.SILENT ()") called at /usr/share/perl5/Mail/IMAPClient.pm line 2
338
        Mail::IMAPClient::store(Mail::IMAPClient=HASH(0x5564453014d8), 668, "FLAGS.SILENT ()") called at /opt/imapsync/imapsync line 6112
        main::sync_flags(HASH(0x55644229d030), "INBOX", 9586, "NonJunk category.Perso", "INBOX", 668, "NonJunk", "") called at /opt/imapsync/imapsync line 6058
        main::sync_flags_fir(HASH(0x55644229d030), "INBOX", 9586, "INBOX", 668, "", HASH(0x556445c26b78), HASH(0x5564454b4358)) called at /opt/imapsync/imapsync line 27
69
        main::single_sync(HASH(0x55644229d030), HASH(0x5564408014d8), HASH(0x55644229d498)) called at /opt/imapsync/imapsync line 1250
reconnecting to os-gravelines-1.webmsg.me, last error: timeout waiting 120s for data from server
Connecting with IO::Socket::SSL PeerAddr os-gravelines-1.webmsg.me PeerPort 993 Proto tcp Timeout 120 Debug 1 SSL_cipher_list DEFAULT:!DH SSL_verify_mode 0 SSL_verifycn
_scheme imap
DEBUG: .../IO/Socket/SSL.pm:3010: new ctx 93889243497008
DEBUG: .../IO/Socket/SSL.pm:762: socket not yet connected
DEBUG: .../IO/Socket/SSL.pm:1177: global error: Undefined SSL object
DEBUG: .../IO/Socket/SSL.pm:1177: global error: Undefined SSL object
DEBUG: .../IO/Socket/SSL.pm:764: socket connected
DEBUG: .../IO/Socket/SSL.pm:787: ssl handshake not started
DEBUG: .../IO/Socket/SSL.pm:829: using SNI with hostname os-gravelines-1.webmsg.me
DEBUG: .../IO/Socket/SSL.pm:880: set socket to non-blocking to enforce timeout=120
DEBUG: .../IO/Socket/SSL.pm:894: call Net::SSLeay::connect
DEBUG: .../IO/Socket/SSL.pm:897: done Net::SSLeay::connect -> 1
DEBUG: .../IO/Socket/SSL.pm:952: ssl handshake done
Connected to os-gravelines-1.webmsg.me
DEBUG: .../IO/Socket/SSL.pm:3059: free ctx 93889080230576 open=93889243497008 93889080230576 93889209037856                                                             
DEBUG: .../IO/Socket/SSL.pm:3070: OK free ctx 93889080230576
Read:   * OK [CAPABILITY IMAP4rev1 LITERAL+ ID ENABLE STARTTLS AUTH=PLAIN SASL-IR] os-gravelines-1.webmsg.me Cyrus IMAP 3.4.2-dirty-Debian-3.4.2-1~bpo11+1 server ready
Sending: 4070 LOGIN mildred@mildred.fr [Redact: Count=4070 Showcredentials=OFF]
Sent 41 bytes
Read:   4070 OK [CAPABILITY IMAP4rev1 LITERAL+ ID ENABLE ACL ANNOTATE-EXPERIMENT-1 BINARY CATENATE CHILDREN CONDSTORE CREATE-SPECIAL-USE ESEARCH ESORT LIST-EXTENDED LIS
T-MYRIGHTS LIST-STATUS MAILBOX-REFERRALS METADATA MOVE MULTIAPPEND NAMESPACE OBJECTID QRESYNC QUOTA RIGHTS=kxten SAVEDATE SEARCH=FUZZY SORT SORT=DISPLAY SPECIAL-USE STA
TUS=SIZE THREAD=ORDEREDSUBJECT THREAD=REFERENCES UIDPLUS UNSELECT URLAUTH URLAUTH=BINARY WITHIN DIGEST=SHA1 LIST-METADATA NO_ATOMIC_RENAME PREVIEW=FUZZY SCAN SORT=MODSE
Q SORT=UID THREAD=REFS X-CREATEDMODSEQ X-REPLICATION XLIST XMOVE LOGINDISABLED XCONVERSATIONS COMPRESS=DEFLATE X-QUOTA=STORAGE X-QUOTA=MESSAGE X-QUOTA=X-ANNOTATION-STOR
AGE X-QUOTA=X-NUM-FOLDERS IDLE] User logged in SESSIONID=<cyrus-1637308599-80986-1-11610858836795431422>
Sending: 4071 SELECT INBOX
Sent 19 bytes
Read:   * 4100 EXISTS
        * 0 RECENT
        * FLAGS (\Answered \Flagged \Draft \Deleted \Seen $Forwarded $label3 $MDNSent $toreply NonJunk Junk)
        * OK [PERMANENTFLAGS (\Answered \Flagged \Draft \Deleted \Seen $Forwarded $label3 $MDNSent $toreply NonJunk Junk \*)] Ok
        * OK [UNSEEN 10] Ok
        * OK [UIDVALIDITY 1628200824] Ok
        * OK [UIDNEXT 4142] Ok
        * OK [HIGHESTMODSEQ 479815] Ok
        * OK [URLMECH INTERNAL] Ok
        * OK [ANNOTATIONS 65536] Ok
        4071 OK [READ-WRITE] Completed
reconnect success(1) on try #1/3
Sending: 4072 UID STORE 668 FLAGS.SILENT ()
Sent 36 bytes
Read:   4072 OK Completed
Host1: size msg INBOX/9586 = 1081199 <> 1081199 = Host2 INBOX/668
Host1: found that msg INBOX/9587 equals Host2 INBOX/669
Host1: flags init msg INBOX/9587 flags( NonJunk category.Perso ) Host2 msg INBOX/669 flags( NonJunk )
Host1: flags filt msg INBOX/9587 flags(  ) Host2 msg INBOX/669 flags( NonJunk )
Host2: flags msg INBOX/669 replacing h2 flags( NonJunk ) with h1 flags(  )
Sending: 4073 UID STORE 669 FLAGS.SILENT ()
Sent 36 bytes
...
```

What seems suspicious is what happens just befor ethe connection stalls:

```
Host2: flags msg INBOX/668 replacing h2 flags( NonJunk ) with h1 flags(  )
Sending: 4069 UID STORE 668 FLAGS.SILENT ()
Sent 36 bytes
```

I'm replacing IMAP flags, probably because I have the following option in my script : `--regexflag 's/.*?(?:([\\\$]\S+\s?)|$)/defined($1)?$1:q()/eg'`

This is because on Host1 (Dovecot), I have too much IMAP flags defined (I have a sieve script that defines flags for anything) and on Host2 (Cyrus) I cannot define as many flags (there is a hard limit of acout 200 flags in Cyrus).

Effectively, on Host2, I have a segfault of Cyrus IMAP

**Comment by gilleslamiral** [comment]
So a reconnection sometimes adds the warning "Undefined SSL object". 
Well, I don't know what to do about this.

**Comment by noxxi** [comment]
Hopefully this should be addressed with IO::Socket::SSL 2.073

**Comment by jauderho** [comment]
I was having trouble with the official Docker image with the version of `$RCSfile: imapsync,v $ $Revision: 2.172 $ $Date: 2021/12/07 18:19:42 $` and was seeing errors like the following:

```
++++ Looping on each one of 27 folders to sync
Folder    1/27 []                            -> []                           
Host1 folder one: Could not select: socket closed while reading data from server; error while reading data from server:  (2x)
Folder    2/27 []                         -> []                        
Host1 folder two: Could not select: error while reading data from server: 
Folder    3/27 []                            -> []                           
Host1 folder three: Could not select: error while reading data from server: 
          
DEBUG: .../IO/Socket/SSL.pm:1201: local error: SSL write error (5)
```

I ended up building my own image using a slightly modified Dockerfile to use Ubuntu 22.04 (beta) and does have IO::Socket::SSL 2.073.

Now with `$RCSfile: imapsync,v $ $Revision: 2.186 $ $Date: 2022/02/22 22:39:50 $`, this seems to be actually copying files over and not seeing the SSL error (so far). Will report back when the job is complete.
